<a href="https://colab.research.google.com/github/RedsquareM/Project1/blob/main/HW2Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
#Загружаем kaggle.jason для удобной загрузки данных из kaggle 
from google.colab import files
uploaded = files.upload()

Saving kaggle.json to kaggle.json


In [3]:
# Загружаем kaggle 
! pip install -q kaggle

In [4]:
# Директория где будет хранится .json файл 
! mkdir ~/.kaggle

In [5]:
#Koпируем наш .json файл с аутентификацией 
! cp kaggle.json ~/.kaggle/

In [6]:
#Проверем если kaggle.json сущeствует 
!ls -lha kaggle.json

-rw-r--r-- 1 root root 66 Oct  7 13:15 kaggle.json


In [7]:
# Меняем разрешения на файл 
! chmod 600 ~/.kaggle/kaggle.json

In [ ]:
#Cмотрим на datsets которые представленны
!kaggle datasets list

In [9]:
# Делаем запрос и загружаем данные из kaggle 
! kaggle competitions download --force riiid-test-answer-prediction

100% 1.29G/1.29G [00:19<00:00, 71.6MB/s]
100% 1.29G/1.29G [00:19<00:00, 70.7MB/s]


In [11]:
#Распокуем данные в .csv
!unzip -o riiid-test-answer-prediction.zip -d ./ 

Archive:  riiid-test-answer-prediction.zip
  inflating: ./example_sample_submission.csv  
  inflating: ./example_test.csv      
  inflating: ./lectures.csv          
  inflating: ./questions.csv         
  inflating: ./riiideducation/__init__.py  
  inflating: ./riiideducation/competition.cpython-37m-x86_64-linux-gnu.so  
  inflating: ./train.csv             


Начнем с EDA (Разведочный анализ данных) чтобы понять, что за данные у нас 
Имеем 5 csv файлов с разными данными 

In [2]:
#Загрузим таблицу из данных train.csv 20000 рядов
#Импортируем библиотеку pandas 
import pandas as pd 
train = pd.read_csv('train.csv', nrows=20000)
display(train)


row_id    timestamp  user_id  content_id  content_type_id  \
0           0            0      115        5692                0   
1           1        56943      115        5716                0   
2           2       118363      115         128                0   
3           3       131167      115        7860                0   
4           4       137965      115        7922                0   
...       ...          ...      ...         ...              ...   
19995   19995  14686301509   220268       10498                0   
19996   19996  14686331029   220268       10681                0   
19997   19997  14686358409   220268       10448                0   
19998   19998  14686410560   220268       10523                0   
19999   19999  14686478609   220268       10614                0   

       task_container_id  user_answer  answered_correctly  \
0                      1            3                   1   
1                      2            2                   1   
2                      0            0                   1   
3                      3            0                   1   
4                      4            1                   1   
...                  ...          ...                 ...   
19995               1312            2                   1   
19996               1313            1                   1   
19997               1314            2                   1   
19998               1315            2                   0   
19999               1316            0                   1   

       prior_question_elapsed_time prior_question_had_explanation  
0                              NaN                            NaN  
1                          37000.0                          False  
2                          55000.0                          False  
3                          19000.0                          False  
4                          11000.0                          False  
...                            ...                            ...  
19995                      18000.0                           True  
19996                      22000.0                           True  
19997                      20000.0                           True  
19998                      19000.0                           True  
19999                      25000.0                           True  

[20000 rows x 10 columns]

In [ ]:
#Зделаем боллее интерактивную таблицу используя функцию google 
#%load_ext google.colab.data_table
#train1 = pd.DataFrame(train) 
#train1
#Убираем
#from google.colab import data_table
#data_table.disable_dataframe_formatter()

Проделаем небольшие операции с другими файлами из дата сэта 


In [3]:
# Прочитаем файл
questions = pd.read_csv('questions.csv')
# Покажем первые 10 строк
questions.head(10)



question_id  bundle_id  correct_answer  part            tags
0            0          0               0     1   51 131 162 38
1            1          1               1     1       131 36 81
2            2          2               0     1  131 101 162 92
3            3          3               0     1  131 149 162 29
4            4          4               3     1    131 5 162 38
5            5          5               2     1  131 149 162 81
6            6          6               2     1    10 94 162 92
7            7          7               0     1   61 110 162 29
8            8          8               3     1   131 13 162 92
9            9          9               3     1       10 164 81

In [4]:
# Посмотрим последние 10 строк в lectures.csv
lectures = pd.read_csv('lectures.csv')
lectures.tail(10)

lecture_id  tag  part           type_of
408       32168  136     4           concept
409       32175   82     4           concept
410       32264  133     6           concept
411       32312  176     2           concept
412       32491  179     5  solving question
413       32535    8     5  solving question
414       32570  113     3  solving question
415       32604   24     6           concept
416       32625  142     2           concept
417       32736   82     3           concept

In [5]:
# Данные из example_test  
example_test = pd.read_csv('example_test.csv')
display(example_test)
# Oчень похоже на дата сет из train.csv 
# prior_group_answers_correct и prior_group_responses, имеют NaN, могут не влиять на данные 

row_id  group_num    timestamp     user_id  content_id  content_type_id  \
0         0          0            0   275030867        5729                0   
1         1          0  13309898705   554169193       12010                0   
2         2          0   4213672059  1720860329         457                0   
3         3          0  62798072960   288641214       13262                0   
4         4          0  10585422061  1728340777        6119                0   
..      ...        ...          ...         ...         ...              ...   
99      104          3  13167339284  1900527744        3004                0   
100     105          3  13167339284  1900527744        3003                0   
101     106          3  64497673060     7792299        7908                0   
102     107          3  62798166743   288641214        9077                0   
103     108          3   2059176357  2018567473       12119                0   

     task_container_id  prior_question_elapsed_time  \
0                    0                          NaN   
1                 4427                      19000.0   
2                  240                      17000.0   
3                  266                      23000.0   
4                  162                      72400.0   
..                 ...                          ...   
99                1179                      24667.0   
100               1179                      24667.0   
101                676                      19000.0   
102                269                      25000.0   
103                591                      20000.0   

    prior_question_had_explanation prior_group_answers_correct  \
0                              NaN                          []   
1                             True                         NaN   
2                             True                         NaN   
3                             True                         NaN   
4                             True                         NaN   
..                             ...                         ...   
99                            True                         NaN   
100                           True                         NaN   
101                           True                         NaN   
102                           True                         NaN   
103                           True                         NaN   

    prior_group_responses  
0                      []  
1                     NaN  
2                     NaN  
3                     NaN  
4                     NaN  
..                    ...  
99                    NaN  
100                   NaN  
101                   NaN  
102                   NaN  
103                   NaN  

[104 rows x 11 columns]

In [6]:
example_sample_submission = pd.read_csv('example_sample_submission.csv')
display(example_sample_submission)
#Из этого датасета не много информации которую можно извлечь

row_id  answered_correctly  group_num
0         0                 0.5          0
1         1                 0.5          0
2         2                 0.5          0
3         3                 0.5          0
4         4                 0.5          0
..      ...                 ...        ...
99      104                 0.5          3
100     105                 0.5          3
101     106                 0.5          3
102     107                 0.5          3
103     108                 0.5          3

[104 rows x 3 columns]

In [7]:
#Посмотрим на случайные n строк 
train.sample(n=100)

row_id    timestamp  user_id  content_id  content_type_id  \
5064     5064   7159800622    24418        4941                0   
18386   18386   1276598456   220268         538                0   
16876   16876   8946299691   166728        5339                0   
16904   16904   9013209889   166728        3976                0   
6487     6487   9317306685    24418        8995                0   
...       ...          ...      ...         ...              ...   
4470     4470   6223673512    24418        6815                0   
19817   19817   9580627009   220268        7951                0   
13157   13157     87204885   138650       10158                0   
5164     5164   7311298104    24418        1152                0   
1474     1474  16570405824    13134       10106                0   

       task_container_id  user_answer  answered_correctly  \
5064                2473            3                   1   
18386                430            1                   1   
16876                 51            1                   0   
16904                 79            0                   1   
6487                3503            0                   0   
...                  ...          ...                 ...   
4470                2086            2                   0   
19817               1187            0                   1   
13157                 75            3                   0   
5164                2559            3                   0   
1474                 701            2                   1   

       prior_question_elapsed_time prior_question_had_explanation  
5064                       16000.0                           True  
18386                      16000.0                           True  
16876                      18000.0                           True  
16904                      23000.0                           True  
6487                       23666.0                           True  
...                            ...                            ...  
4470                       30000.0                           True  
19817                      21000.0                           True  
13157                      44250.0                           True  
5164                       20000.0                           True  
1474                       29666.0                           True  

[100 rows x 10 columns]

In [ ]:
#Получим описание по численным данным из train.csv
train.describe()

row_id     timestamp        user_id    content_id  \
count  20000.000000  2.000000e+04   20000.000000  20000.000000   
mean    9999.500000  5.707366e+09   90272.814350   5416.631600   
std     5773.647028  5.151766e+09   69644.353597   4025.878106   
min        0.000000  0.000000e+00     115.000000      0.000000   
25%     4999.750000  1.535160e+09   24418.000000   2081.000000   
50%     9999.500000  4.578822e+09   91216.000000   5263.000000   
75%    14999.250000  8.577729e+09  141455.000000   7744.250000   
max    19999.000000  4.375822e+10  220268.000000  32736.000000   

       content_type_id  task_container_id   user_answer  answered_correctly  \
count      20000.00000       20000.000000  20000.000000        20000.000000   
mean           0.02075        1158.101250      1.390850            0.649900   
std            0.14255        1265.240323      1.196728            0.518694   
min            0.00000           0.000000     -1.000000           -1.000000   
25%            0.00000         251.000000      0.000000            0.000000   
50%            0.00000         715.000000      1.000000            1.000000   
75%            0.00000        1405.250000      3.000000            1.000000   
max            1.00000        4889.000000      3.000000            1.000000   

       prior_question_elapsed_time  
count                 19542.000000  
mean                  24261.523744  
std                   15499.427882  
min                     333.000000  
25%                   16000.000000  
50%                   21000.000000  
75%                   28000.000000  
max                  300000.000000

In [ ]:
#посмотрим какие есть признаки, какой у них тип данных в таблице и сколько есть заполненных ячеек 
train.info()
#10 колонок, но видимо row_id и timestamp не совсем важны

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 10 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   row_id                          20000 non-null  int64  
 1   timestamp                       20000 non-null  int64  
 2   user_id                         20000 non-null  int64  
 3   content_id                      20000 non-null  int64  
 4   content_type_id                 20000 non-null  int64  
 5   task_container_id               20000 non-null  int64  
 6   user_answer                     20000 non-null  int64  
 7   answered_correctly              20000 non-null  int64  
 8   prior_question_elapsed_time     19542 non-null  float64
 9   prior_question_had_explanation  19957 non-null  object 
dtypes: float64(1), int64(8), object(1)
memory usage: 1.5+ MB


In [30]:
#Посмотрим, в каких столбцах есть пропуски, так как эти данные могут представить проблемы при анализе
#Как видно есть пропуски в prior_questions_elapsed_time и prior_question_had_explanation 

train.isnull().sum()

row_id                              0
timestamp                           0
user_id                             0
content_id                          0
content_type_id                     0
task_container_id                   0
user_answer                         0
answered_correctly                  0
prior_question_elapsed_time       458
prior_question_had_explanation     43
dtype: int64

In [31]:
#какая доля пропусков содержится в колонке 
train.isnull().mean() 
#2% данных в prior_question_elapased_time пропущенно 

row_id                            0.00000
timestamp                         0.00000
user_id                           0.00000
content_id                        0.00000
content_type_id                   0.00000
task_container_id                 0.00000
user_answer                       0.00000
answered_correctly                0.00000
prior_question_elapsed_time       0.02290
prior_question_had_explanation    0.00215
dtype: float64

In [32]:
# Удалим те строки, в которых пропущено значение в колонке
train.dropna(axis=0, inplace=True)
train.isnull().sum()

row_id                            0
timestamp                         0
user_id                           0
content_id                        0
content_type_id                   0
task_container_id                 0
user_answer                       0
answered_correctly                0
prior_question_elapsed_time       0
prior_question_had_explanation    0
dtype: int64

In [ ]:
#Посмотрим по столбам и строкам до 2000 строк 
#Видим, что время варируется в зависимости от ответа 
train.loc[:2000, 'user_answer':'prior_question_had_explanation'] 

user_answer  answered_correctly  prior_question_elapsed_time  \
1               2                   1                      37000.0   
2               0                   1                      55000.0   
3               0                   1                      19000.0   
4               1                   1                      11000.0   
5               2                   1                       5000.0   
...           ...                 ...                          ...   
1996            1                   1                      18000.0   
1997            0                   0                      14000.0   
1998            0                   1                      16000.0   
1999            1                   1                      18000.0   
2000            0                   1                      15000.0   

     prior_question_had_explanation  
1                             False  
2                             False  
3                             False  
4                             False  
5                             False  
...                             ...  
1996                           True  
1997                           True  
1998                           True  
1999                           True  
2000                           True  

[1970 rows x 4 columns]

In [14]:
# Посмотрим какие ответы есть
train['answered_correctly'].unique()
# Возможные ответы, либо 1 или 0, -1, -1 as null, for lectures.

array([ 1,  0, -1])

Теперь посмотрим какие ответы: answered_correctly либо (0 and 1), and relationship with prior_question_had_explanation is either False and True 
Из результатов видно что самое большое правильных ответов (12943) когда было представлено (prior_queston_had_explanation). То есть если предположить, когда дано объяснение скорей всего студент ответит правильно. 
Испробуем различные варианты и комбинации ответов и отношение с "prior_question_had_explanation" колонкой 

In [8]:
train[(train['answered_correctly'] == 0) & (train['prior_question_had_explanation'] == False)].value_counts()

row_id  timestamp   user_id  content_id  content_type_id  task_container_id  user_answer  answered_correctly  prior_question_elapsed_time  prior_question_had_explanation
10      255381      115      152         0                10                 2            0                   17000.0                      False                             1
13084   986559      138650   7219        0                15                 0            0                   27000.0                      False                             1
14223   421441      141455   6370        0                13                 0            0                   12000.0                      False                             1
14221   392039      141455   6116        0                11                 1            0                   7000.0                       False                             1
14219   366707      141455   4120        0                9                  3            0                   22000.0             

In [ ]:
train[(train['answered_correctly'] == 1) & (train['prior_question_had_explanation'] == False)].value_counts()

row_id  timestamp   user_id  content_id  content_type_id  task_container_id  user_answer  answered_correctly  prior_question_elapsed_time  prior_question_had_explanation
1       56943       115      5716        0                2                  2            1                   37000.0                      False                             1
13080   678556      138650   6910        0                14                 3            1                   20000.0                      False                             1
14182   310471      140969   6173        0                12                 2            1                   8000.0                       False                             1
14178   263078      140969   4492        0                8                  2            1                   8666.0                       False                             1
14175   235214      140969   2594        0                7                  2            1                   10333.0             

In [ ]:
train[(train['answered_correctly'] == 0) & (train['prior_question_had_explanation'] == True)].value_counts()

row_id  timestamp    user_id  content_id  content_type_id  task_container_id  user_answer  answered_correctly  prior_question_elapsed_time  prior_question_had_explanation
36      1359412      115      7897        0                36                 3            0                   17000.0                      True                              1
14129   26139169124  138650   8833        0                810                1            0                   7000.0                       True                              1
13764   4709775023   138650   7219        0                471                3            0                   14000.0                      True                              1
13759   4610250252   138650   5092        0                467                2            0                   22000.0                      True                              1
13758   4610212981   138650   4065        0                466                0            0                   12000.0       

In [ ]:
train[(train['answered_correctly'] == 1) & (train['prior_question_had_explanation'] == True)].value_counts()

row_id  timestamp    user_id  content_id  content_type_id  task_container_id  user_answer  answered_correctly  prior_question_elapsed_time  prior_question_had_explanation
33      1252621      115      7948        0                33                 0            1                   20000.0                      True                              1
13194   132366955    138650   6772        0                96                 0            1                   26500.0                      True                              1
13178   131751090    138650   10685       0                92                 0            1                   17000.0                      True                              1
13179   131907920    138650   6765        0                93                 1            1                   17000.0                      True                              1
13180   131907920    138650   6767        0                93                 1            1                   17000.0       

In [15]:
# Ответы пользователя варируется от 5 возможных, скорей всего вопросы которые были использованы - это вопросы с несколькими вариантами ответов
train['user_answer'].unique()

array([ 3,  2,  0,  1, -1])

In [19]:
# Посмотрим какие ответы имеют наибольшое значение для правильных ответов 
# Видим что 
questions['correct_answer'].value_counts()

0    3716
3    3544
1    3478
2    2785
Name: correct_answer, dtype: int64

In [9]:
#Построим новую дата фрэйм с более важными колонками 
col_list = ['content_type_id','user_answer', 'user_id', 'answered_correctly', 'prior_question_had_explanation']
train_reduced = pd.read_csv('train.csv', usecols=col_list)


In [10]:
#Видим что у нас более 100М рекордов в таблице 
display(train_reduced)
train_reduced.shape

user_id  content_type_id  user_answer  answered_correctly  \
0                 115                0            3                   1   
1                 115                0            2                   1   
2                 115                0            0                   1   
3                 115                0            0                   1   
4                 115                0            1                   1   
...               ...              ...          ...                 ...   
101230327  2147482888                0            0                   1   
101230328  2147482888                0            3                   1   
101230329  2147482888                0            3                   1   
101230330  2147482888                0            1                   0   
101230331  2147482888                0            3                   1   

          prior_question_had_explanation  
0                                    NaN  
1                                  False  
2                                  False  
3                                  False  
4                                  False  
...                                  ...  
101230327                           True  
101230328                           True  
101230329                           True  
101230330                           True  
101230331                           True  

[101230332 rows x 5 columns]

(101230332, 5)

Найдем разные счетчики таблицы

In [36]:
#1.Посмотрим сколько людей участвовали в эксперементе (user_id)
train_reduced['user_id'].value_counts().max()

17917

In [9]:
#2 Non unique values in content type 
train_reduced['content_type_id'].nunique()

2

In [11]:
#3 unique values in answered_correctly 
train_reduced['answered_correctly'].unique()

array([ 1,  0, -1])

In [12]:
# Полный счет ответов в дата сет 
train_reduced['user_answer'].value_counts()

 0    28186489
 1    26990007
 3    26084784
 2    18010020
-1     1959032
Name: user_answer, dtype: int64

In [11]:
# Сделаем datafrime из 2-x колонках
col_list = ['content_id','user_answer']
train_content_id = pd.read_csv('train.csv', usecols=col_list)

In [12]:
display(train_content_id)

content_id  user_answer
0                5692            3
1                5716            2
2                 128            0
3                7860            0
4                7922            1
...               ...          ...
101230327        3586            0
101230328        6341            3
101230329        4212            3
101230330        6343            1
101230331        7995            3

[101230332 rows x 2 columns]

In [50]:
#Посмотрим на distrubition of content_id
train_content_id['content_id'].value_counts()
#Moжно понять что самое большая частота распределения под content_id = 6116

6116     213605
6173     202106
4120     199372
175      195861
7876     190170
          ...  
10033         1
1486          1
21169         1
1485          1
10005         1
Name: content_id, Length: 13782, dtype: int64

In [21]:
#Посмотрим среднее время в миллисекундах между взаимодействием с пользователем и первым завершением события от этого пользователя.
col_list = ['user_id', 'timestamp']
train_time = pd.read_csv('train.csv', usecols=col_list)

In [22]:
display(train_time)

timestamp     user_id
0                  0         115
1              56943         115
2             118363         115
3             131167         115
4             137965         115
...              ...         ...
101230327  428564420  2147482888
101230328  428585000  2147482888
101230329  428613475  2147482888
101230330  428649406  2147482888
101230331  428692118  2147482888

[101230332 rows x 2 columns]

In [33]:
millis = train_time['timestamp'].mean().astype(int)
seconds = (millis/1000)%60
print(round(seconds,1))
# Вывод что в среднем занимает 3.7 секунды между тем как пользователь совершит действие при взаимодействие с системой 

3.7


In [46]:
#Посмотрим кто из пользователей занял самое долгое время перед действием
#Как видно это был user_id = 705741139 
train_time.groupby(['user_id'])['timestamp'].max().sort_values(ascending=False)

user_id
705741139     87425772049
626308830     87193355096
1887984732    86764051932
1857133602    86735591752
881103868     86541352191
                 ...     
692663988               0
1181377042              0
121434886               0
1869558407              0
142134372               0
Name: timestamp, Length: 393656, dtype: int64

Teперь попрубем провести более глубокий анализ данных, после того как уже имеем какое-то представаление о данных
Испробуем разные фильтры, когда ответ =0  и Тrue

In [ ]:
train_reduced[(train_reduced['answered_correctly'] == 0) & (train_reduced['prior_question_had_explanation'] == True)].value_counts()

content_type_id  user_answer  answered_correctly  prior_question_had_explanation
0                0            0                   True                              8294247
                 1            0                   True                              8146156
                 3            0                   True                              7708925
                 2            0                   True                              5160944
dtype: int64

In [ ]:
# Статические данные 
train_reduced.user_answer.describe()

count    1.012303e+08
mean     1.376123e+00
std      1.192896e+00
min     -1.000000e+00
25%      0.000000e+00
50%      1.000000e+00
75%      3.000000e+00
max      3.000000e+00
Name: user_answer, dtype: float64

In [ ]:
train_reduced.user_answer.value_counts().head(1).index[0]

0

In [ ]:
train_reduced.answered_correctly.value_counts().head(1).index[0]

1

In [ ]:
#определите количество различных типов вопросов которые были представленны из lectures.csv
lectures['type_of'].unique()

array(['concept', 'solving question', 'intention', 'starter'],
      dtype=object)

In [ ]:
#Какой тип вопросов был задан больше всего и меньше всего
lectures['type_of'].value_counts()
#Самое большое вопросов было задано по типу: concept, и меньше всего по starter

concept             222
solving question    186
intention             7
starter               3
Name: type_of, dtype: int64

In [ ]:
# Сколько частей было в lectures и их количество
lectures['part'].value_counts()

5    143
6     83
2     56
1     54
7     32
4     31
3     19
Name: part, dtype: int64

In [ ]:
# Как part совпадает с part из lectures
questions['part'].value_counts()

5    5511
2    1647
3    1562
4    1439
6    1212
7    1160
1     992
Name: part, dtype: int64

In [ ]:
#Получим описание по численным данным из questions.csv
questions.describe()

question_id     bundle_id  correct_answer          part
count  13523.00000  13523.000000    13523.000000  13523.000000
mean    6761.00000   6760.510907        1.455298      4.264956
std     3903.89818   3903.857783        1.149707      1.652553
min        0.00000      0.000000        0.000000      1.000000
25%     3380.50000   3379.500000        0.000000      3.000000
50%     6761.00000   6761.000000        1.000000      5.000000
75%    10141.50000  10140.000000        3.000000      5.000000
max    13522.00000  13522.000000        3.000000      7.000000

In [ ]:
questions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   question_id     13523 non-null  int64 
 1   bundle_id       13523 non-null  int64 
 2   correct_answer  13523 non-null  int64 
 3   part            13523 non-null  int64 
 4   tags            13522 non-null  object
dtypes: int64(4), object(1)
memory usage: 528.4+ KB


In [ ]:
#Получим описание по численным данным из lectures.csv
lectures.describe()

lecture_id         tag        part
count    418.000000  418.000000  418.000000
mean   16983.401914   94.480861    4.267943
std     9426.164660   53.586487    1.872424
min       89.000000    0.000000    1.000000
25%     9026.250000   50.250000    2.000000
50%    17161.500000   94.500000    5.000000
75%    24906.250000  140.000000    6.000000
max    32736.000000  187.000000    7.000000

In [ ]:
lectures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 418 entries, 0 to 417
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   lecture_id  418 non-null    int64 
 1   tag         418 non-null    int64 
 2   part        418 non-null    int64 
 3   type_of     418 non-null    object
dtypes: int64(3), object(1)
memory usage: 13.2+ KB


In [ ]:
display(lectures)

lecture_id  tag  part           type_of
0            89  159     5           concept
1           100   70     1           concept
2           185   45     6           concept
3           192   79     5  solving question
4           317  156     5  solving question
..          ...  ...   ...               ...
413       32535    8     5  solving question
414       32570  113     3  solving question
415       32604   24     6           concept
416       32625  142     2           concept
417       32736   82     3           concept

[418 rows x 4 columns]

In [ ]:
# Попрубуем посмотреть корреляцию по пеарсону, (-) означает негативная коррелеция, (+) позитивная коррелеция 
# По результатам видно что нет корреляции так как значения близки к нулю
train.corr(method='pearson')

row_id  timestamp   user_id  content_id  \
row_id                       1.000000  -0.291129  0.956083    0.004416   
timestamp                   -0.291129   1.000000 -0.347321    0.070780   
user_id                      0.956083  -0.347321  1.000000    0.007921   
content_id                   0.004416   0.070780  0.007921    1.000000   
content_type_id             -0.019729  -0.011115 -0.015346    0.425648   
task_container_id           -0.274931   0.509214 -0.450894   -0.040962   
user_answer                  0.020211  -0.003254  0.014991   -0.120193   
answered_correctly           0.008271  -0.006660  0.011610   -0.196623   
prior_question_elapsed_time  0.082344  -0.115658  0.080406    0.164397   

                             content_type_id  task_container_id  user_answer  \
row_id                             -0.019729          -0.274931     0.020211   
timestamp                          -0.011115           0.509214    -0.003254   
user_id                            -0.015346          -0.450894     0.014991   
content_id                          0.425648          -0.040962    -0.120193   
content_type_id                     1.000000           0.007498    -0.290824   
task_container_id                   0.007498           1.000000     0.006063   
user_answer                        -0.290824           0.006063     1.000000   
answered_correctly                 -0.463041           0.014794     0.127334   
prior_question_elapsed_time              NaN          -0.074474     0.013476   

                             answered_correctly  prior_question_elapsed_time  
row_id                                 0.008271                     0.082344  
timestamp                             -0.006660                    -0.115658  
user_id                                0.011610                     0.080406  
content_id                            -0.196623                     0.164397  
content_type_id                       -0.463041                          NaN  
task_container_id                      0.014794                    -0.074474  
user_answer                            0.127334                     0.013476  
answered_correctly                     1.000000                     0.021175  
prior_question_elapsed_time            0.021175                     1.000000

In [6]:
# ковариация показывает как связаны две переменные, (+) - прямая зависимость, (-) обратная зависимость 
train.cov()
# Самая большая зависимость наблюдается между task_container_id и answered_correctly как и ожидалось 

row_id     timestamp       user_id  \
row_id                       3.333500e+07 -8.659467e+12  3.844428e+08   
timestamp                   -8.659467e+12  2.654069e+19 -1.246158e+14   
user_id                      3.844428e+08 -1.246158e+14  4.850336e+09   
content_id                   1.026467e+05  1.467998e+12  2.220847e+06   
content_type_id             -1.623794e+01 -8.162702e+06 -1.523493e+02   
task_container_id           -2.008383e+06  3.319169e+12 -3.973135e+07   
user_answer                  1.396476e+02 -2.006315e+07  1.249466e+03   
answered_correctly           2.477064e+01 -1.779705e+07  4.193912e+02   
prior_question_elapsed_time  7.368642e+06 -9.260833e+12  8.683346e+07   

                               content_id  content_type_id  task_container_id  \
row_id                       1.026467e+05    -1.623794e+01      -2.008383e+06   
timestamp                    1.467998e+12    -8.162702e+06       3.319169e+12   
user_id                      2.220847e+06    -1.523493e+02      -3.973135e+07   
content_id                   1.620769e+07     2.442743e+02      -2.086501e+05   
content_type_id              2.442743e+02     2.032045e-02       1.352267e+00   
task_container_id           -2.086501e+05     1.352267e+00       1.600833e+06   
user_answer                 -5.790742e+02    -4.961262e-02       9.180835e+00   
answered_correctly          -4.105883e+02    -3.423714e-02       9.708783e+00   
prior_question_elapsed_time  8.676415e+06     0.000000e+00      -1.463451e+06   

                              user_answer  answered_correctly  \
row_id                       1.396476e+02        2.477064e+01   
timestamp                   -2.006315e+07       -1.779705e+07   
user_id                      1.249466e+03        4.193912e+02   
content_id                  -5.790742e+02       -4.105883e+02   
content_type_id             -4.961262e-02       -3.423714e-02   
task_container_id            9.180835e+00        9.708783e+00   
user_answer                  1.432158e+00        7.904054e-02   
answered_correctly           7.904054e-02        2.690434e-01   
prior_question_elapsed_time  2.416010e+02        1.524824e+02   

                             prior_question_elapsed_time  
row_id                                      7.368642e+06  
timestamp                                  -9.260833e+12  
user_id                                     8.683346e+07  
content_id                                  8.676415e+06  
content_type_id                             0.000000e+00  
task_container_id                          -1.463451e+06  
user_answer                                 2.416010e+02  
answered_correctly                          1.524824e+02  
prior_question_elapsed_time                 2.402323e+08

In [13]:
questions.cov()
#Наблюдается большая зависимость между correct_answer и question_id и part 

question_id     bundle_id  correct_answer         part
question_id     1.524042e+07  1.524026e+07      133.667653  1411.143396
bundle_id       1.524026e+07  1.524011e+07      133.665533  1410.720634
correct_answer  1.336677e+02  1.336655e+02        1.321826     0.041389
part            1.411143e+03  1.410721e+03        0.041389     2.730932

In [14]:
lectures.cov()
#Taкже имеем зависимость между tag и lecture_id

lecture_id           tag        part
lecture_id  8.885258e+07  14112.285888  494.208599
tag         1.411229e+04   2871.511623   -4.121958
part        4.942086e+02     -4.121958    3.505972

Итак имеем более детальное представление о данных, теперь можно соединить некоторые dataframes и проанализировать непосредственно успеваемость 
Mожно сравнить correct_answer со столбцом train user_answer, чтобы проверить, был ли пользователь прав.

In [13]:
# Соединим по правилу join две dataframes 
df_lectures = pd.read_csv('lectures.csv')
df_questions = pd.read_csv('questions.csv')
df_all_rows = pd.concat([df_lectures, df_questions], ignore_index=True)
df_combined = pd.merge(df_lectures, df_questions, how='inner', left_on = 'lecture_id', right_on = 'question_id')
df_combined

lecture_id  tag  part_x           type_of  question_id  bundle_id  \
0            89  159       5           concept           89         89   
1           100   70       1           concept          100        100   
2           185   45       6           concept          185        185   
3           192   79       5  solving question          192        192   
4           317  156       5  solving question          317        317   
..          ...  ...     ...               ...          ...        ...   
153       13336  113       4           concept        13336      13336   
154       13379  124       5           concept        13379      13379   
155       13384   61       1  solving question        13384      13384   
156       13471   47       5  solving question        13471      13471   
157       13487   98       7  solving question        13487      13487   

     correct_answer  part_y            tags  
0                 1       1       131 36 38  
1                 0       1        131 5 81  
2                 0       1      131 111 81  
3                 1       1      131 104 38  
4                 0       2  143 141 38 102  
..              ...     ...             ...  
153               3       5               4  
154               1       5             116  
155               3       5              73  
156               3       5              96  
157               1       5              96  

[158 rows x 9 columns]

In [14]:
df_combined.describe(include='all')

lecture_id         tag      part_x  type_of   question_id  \
count     158.000000  158.000000  158.000000      158    158.000000   
unique           NaN         NaN         NaN        4           NaN   
top              NaN         NaN         NaN  concept           NaN   
freq             NaN         NaN         NaN       82           NaN   
mean     6828.772152   93.943038    4.234177      NaN   6828.772152   
std      3968.835081   53.890340    1.935810      NaN   3968.835081   
min        89.000000    0.000000    1.000000      NaN     89.000000   
25%      3347.500000   49.250000    2.000000      NaN   3347.500000   
50%      7009.000000   93.500000    5.000000      NaN   7009.000000   
75%     10049.000000  140.000000    6.000000      NaN  10049.000000   
max     13487.000000  187.000000    7.000000      NaN  13487.000000   

           bundle_id  correct_answer      part_y tags  
count     158.000000      158.000000  158.000000  158  
unique           NaN             NaN         NaN   99  
top              NaN             NaN         NaN    8  
freq             NaN             NaN         NaN   13  
mean     6828.348101        1.462025    4.227848  NaN  
std      3968.739768        1.132249    1.666067  NaN  
min        89.000000        0.000000    1.000000  NaN  
25%      3347.500000        0.250000    3.000000  NaN  
50%      7006.000000        1.000000    5.000000  NaN  
75%     10047.000000        3.000000    5.000000  NaN  
max     13487.000000        3.000000    7.000000  NaN

In [ ]:
#Зделаем боллее интерактивную таблицу используя функцию google 
%load_ext google.colab.data_table
df_combined = pd.DataFrame(df_combined) 
#train1
#Убираем
#from google.colab import data_table
#data_table.disable_dataframe_formatter()

Проделаем разные фильтрации, посмотреть какие типы вопросы были заданы, и ответы которые были получены 

In [ ]:
df_combined[(df_combined['type_of'] == 'concept') & (df_combined['correct_answer'] == 0)].value_counts()

lecture_id  tag  part_x  type_of  question_id  bundle_id  correct_answer  part_y  tags            
100         70   1       concept  100          100        0               1       131 5 81            1
8138        151  6       concept  8138         8134       0               7       98 84 97 135 162    1
12638       96   5       concept  12638        12638      0               1       131 104 38          1
12354       49   6       concept  12354        12354      0               3       106 38 29           1
11873       152  6       concept  11873        11873      0               2       90 100 92 29        1
10688       39   7       concept  10688        10688      0               2       62 137 142 81 92    1
10217       19   7       concept  10217        10216      0               6       96                  1
10180       33   6       concept  10180        10180      0               6       53                  1
10052       0    7       concept  10052        10050      0          

In [ ]:
df_combined[(df_combined['type_of'] == 'solving question') & (df_combined['correct_answer'] == 1)].value_counts('correct_answer')


correct_answer
1    19
dtype: int64

In [ ]:
df_combined[(df_combined['type_of'] == 'concept') & (df_combined['correct_answer'] == 1)].value_counts('correct_answer')

correct_answer
1    25
dtype: int64

In [ ]:
df_combined[['type_of','correct_answer']]

In [ ]:
df_combined.groupby('correct_answer').tag.mean()

correct_answer
0     92.575000
1    100.413043
2     85.516129
3     94.390244
Name: tag, dtype: float64

In [86]:
df_combined.groupby('correct_answer').part_y.describe()

count      mean       std  min  25%  50%  75%  max
correct_answer                                                    
0                40.0  3.900000  1.891953  1.0  2.0  5.0  5.0  7.0
1                46.0  3.826087  1.637130  1.0  3.0  4.0  5.0  6.0
2                31.0  4.774194  1.606439  1.0  4.0  5.0  5.5  7.0
3                41.0  4.585366  1.341186  1.0  5.0  5.0  5.0  7.0

Соединим теперь questions с train

In [77]:
# Соединим по правилу join две dataframes 
df_train = pd.read_csv('train.csv', nrows=100)
df_questions = pd.read_csv('questions.csv', nrows=100)
df_all_rows = pd.concat([df_train, df_questions], ignore_index=True)
df_combined1 = pd.merge(df_train, df_questions, how='inner', left_on = 'user_answer', right_on = 'question_id')
df_combined1

row_id  timestamp  user_id  content_id  content_type_id  \
0        0          0      115        5692                0   
1        7     194190      115          50                0   
2       19     468511      115        7927                0   
3       21     510583      115        7984                0   
4       23     557677      115         185                0   
..     ...        ...      ...         ...              ...   
94      75     571323      124        7219                0   
95      76          0     2746        5273                0   
96      82     134341     2746         405                0   
97      97      39828     5382        3944                0   
98      99     153727     5382        5844                0   

    task_container_id  user_answer  answered_correctly  \
0                   1            3                   1   
1                   7            3                   1   
2                  19            3                   1   
3                  21            3                   0   
4                  23            3                   0   
..                ...          ...                 ...   
94                 15            1                   0   
95                  0            1                   0   
96                  6            1                   1   
97                  1            1                   0   
98                  3            1                   0   

    prior_question_elapsed_time prior_question_had_explanation  question_id  \
0                           NaN                            NaN            3   
1                       17000.0                          False            3   
2                       17000.0                          False            3   
3                       20000.0                          False            3   
4                       21000.0                          False            3   
..                          ...                            ...          ...   
94                       6500.0                          False            1   
95                          NaN                            NaN            1   
96                      16000.0                          False            1   
97                      24000.0                          False            1   
98                      88000.0                          False            1   

    bundle_id  correct_answer  part            tags  
0           3               0     1  131 149 162 29  
1           3               0     1  131 149 162 29  
2           3               0     1  131 149 162 29  
3           3               0     1  131 149 162 29  
4           3               0     1  131 149 162 29  
..        ...             ...   ...             ...  
94          1               1     1       131 36 81  
95          1               1     1       131 36 81  
96          1               1     1       131 36 81  
97          1               1     1       131 36 81  
98          1               1     1       131 36 81  

[99 rows x 15 columns]

In [78]:
#Уберем ненужные колонки 
columns = ['row_id', 'user_id', 'content_id', 'task_container_id', 'bundle_id', 'tags', 'part', 'timestamp', 'content_type_id', 'prior_question_elapsed_time', 'prior_question_had_explanation']
df_combined1.drop(columns, inplace=True, axis=1)



In [79]:
# Возьмем небольшой sample из population и проработаем главную тему успеваемости и сравним правильность ответов 
display(df_combined1)

user_answer  answered_correctly  question_id  correct_answer
0             3                   1            3               0
1             3                   1            3               0
2             3                   1            3               0
3             3                   0            3               0
4             3                   0            3               0
..          ...                 ...          ...             ...
94            1                   0            1               1
95            1                   0            1               1
96            1                   1            1               1
97            1                   0            1               1
98            1                   0            1               1

[99 rows x 4 columns]

In [61]:
df_combined1[(df_combined1['answered_correctly'] == 1) & (df_combined1['correct_answer'] == 1)].value_counts()

user_answer  answered_correctly  question_id  correct_answer
1            1                   1            1                 9
dtype: int64

In [60]:
df_combined1[(df_combined1['answered_correctly'] == 0) & (df_combined1['correct_answer'] == 0)].value_counts()

user_answer  answered_correctly  question_id  correct_answer
3            0                   3            0                 17
0            0                   0            0                 10
2            0                   2            0                  8
dtype: int64

In [62]:
df_combined1[(df_combined1['answered_correctly'] == 1) & (df_combined1['correct_answer'] == 0)].value_counts()

user_answer  answered_correctly  question_id  correct_answer
0            1                   0            0                 17
3            1                   3            0                 15
2            1                   2            0                 11
dtype: int64

In [63]:
df_combined1[(df_combined1['answered_correctly'] == 0) & (df_combined1['correct_answer'] == 1)].value_counts()

user_answer  answered_correctly  question_id  correct_answer
1            0                   1            1                 12
dtype: int64

In [ ]:
#import pandas as pd 
#import matplotlib.pyplot as plt 
#import seaborn as sns

#plt.figure(figsize=(12,8))
#sns.countplot(df_combined1['user_answer'], hue=df_combined1['answered_correctly'])
#plt.title('User_Answer vs Answered_Correctly', fontsize = 25)
#plt.show()

Краткие ВЫВОДЫ ПО ДАТЕ

1.Используя определенные колонки смогли увидеть количество рядов в файле train.csv - более 100М (сделали EDA analysis) 

2.Из lectures выяснили что самое большое вопросов было задано по типу: concept, и меньше всего по starter

3.Дата имела очень мало пропусков где-то около 2% для статистики пропуски были удаленны

4.Выяснили разные count метрики, например correct_answers:
0    3716
3    3544
1    3478
2    2785

5.Подсчитали среднее время на ответ около 3.7 секунды 

6.Посмотрели как prior_explanation влияет на ответ

7.Выяснили что из sample, когда user давал ответ 1 или 2 больше всего совпадало правильных ответов 